### PR TITLE
sql: collect cluster-wide traces as part of job execution details

### PR DESCRIPTION
**Pull request by adityamaru**
This change includes the collection of cluster-wide, active tracing
spans to a job's execution details. The traces are stored in a zip
with a text and jaegar json file per node, that contain the active
tracing spans associated with the current execution of the job.

Once https://github.com/cockroachdb/cockroach/pull/106879
and https://github.com/cockroachdb/cockroach/pull/107210
merge a trace.zip will be generated everytime a user requests
new execution details from the job details page. This zip will
be downloadable from the job details page itself.

Informs: https://github.com/cockroachdb/cockroach/issues/102794
Release note: None

### CODE REVIEW

File: pkg/sql/jobs_profiler_execution_details.go [hunk 1]
```diff
@@ -244,7 +248,45 @@ func (e *executionDetailsBuilder) addDistSQLDiagram(ctx context.Context) {
...
+		progressBytes, exists, err := jobInfo.GetLegacyProgress(ctx)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return errors.Newf("progress not found found job %d", e.jobID)
+		}
+		var progress jobspb.Progress
+		if err := protoutil.Unmarshal(progressBytes, &progress); err != nil {
+			return errors.New("failed to unmarshal progress bytes")
```
Comment (stevendanna) [inline pkg/sql/jobs_profiler_execution_details.go:273]: Wrapf?

File: pkg/sql/jobs_profiler_execution_details.go [hunk 2]
```diff
@@ -244,7 +248,45 @@ func (e *executionDetailsBuilder) addDistSQLDiagram(ctx context.Context) {
...
 	}
 }
+
+// addClusterWideTraces generates and persists a `trace.<timestamp>.zip` file
+// that captures the active tracing spans of a job on all nodes in the cluster.
+func (e *executionDetailsBuilder) addClusterWideTraces(ctx context.Context) {
+	z := zipper.MakeInternalExecutorInflightTraceZipper(e.db.Executor())
+
+	var traceID tracingpb.TraceID
+	if err := e.db.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
```
Comment (stevendanna) [inline pkg/sql/jobs_profiler_execution_details.go:262]: Might be nice to extract this to a function

File: pkg/sql/jobs_profiler_execution_details_test.go [hunk 3]
```diff
@@ -220,6 +223,81 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
...
+					return nil
+				},
+			}
+		}, jobs.UsesTenantCostControl)
+		var importJobID int
+		runner.QueryRow(t, `IMPORT INTO t CSV DATA ('nodelocal://1/foo') WITH DETACHED`).Scan(&importJobID)
+		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
+		runner.Exec(t, `SELECT crdb_internal.request_job_execution_details($1)`, importJobID)
+		trace := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "resumer-trace")
+		require.True(t, strings.Contains(string(trace), "should see this"))
```
Comment (stevendanna) [inline pkg/sql/jobs_profiler_execution_details_test.go:243]: If you fancy, there is a.`require.Contains`

File: pkg/sql/jobs_profiler_execution_details_test.go [hunk 4]
```diff
@@ -220,6 +223,81 @@ func TestReadWriteProfilerExecutionDetails(t *testing.T) {
...
+		activeTraces := checkExecutionDetails(t, s, jobspb.JobID(importJobID), "trace")
+		continueCh <- struct{}{}
+		jobutils.WaitForJobToSucceed(t, runner, jobspb.JobID(importJobID))
+		unzip, err := zip.NewReader(bytes.NewReader(activeTraces), int64(len(activeTraces)))
+		require.NoError(t, err)
+
+		// Make sure the bundle contains the expected list of files.
+		var files []string
+		for _, f := range unzip.File {
+			t.Logf("found file: %s", f.Name)
```
Comment (stevendanna) [inline pkg/sql/jobs_profiler_execution_details_test.go:275]: Definitely useful?

File: pkg/sql/jobs_profiler_execution_details.go [hunk 5]
```diff
@@ -244,7 +248,45 @@ func (e *executionDetailsBuilder) addDistSQLDiagram(ctx context.Context) {
...
+		jobInfo := jobs.InfoStorageForJob(txn, e.jobID)
+		progressBytes, exists, err := jobInfo.GetLegacyProgress(ctx)
+		if err != nil {
+			return err
+		}
+		if !exists {
+			return errors.Newf("progress not found found job %d", e.jobID)
```
Comment (stevendanna) [inline pkg/sql/jobs_profiler_execution_details.go:269]: perhaps "progress not found for job %d"

I'll note we logging this immediately below and also putting the jobID in that log message. so perhaps we could elide it here completely.

**Review by stevendanna — COMMENTED** [review]
Cool. Looks like we have an existing test to update.

Also, it might be good to put an example of how to use  these improvements in the commit message/PR message so it is easier to play with.

File: pkg/sql/jobs_profiler_execution_details.go (hunk 2 above)
Comment (adityamaru) [reply pkg/sql/jobs_profiler_execution_details.go:262]: good idea, done

File: pkg/sql/jobs_profiler_execution_details.go (hunk 5 above)
Comment (adityamaru) [reply pkg/sql/jobs_profiler_execution_details.go:269]: yup, deleted the jobID from the first error message

File: pkg/sql/jobs_profiler_execution_details_test.go (hunk 4 above)
Comment (adityamaru) [reply pkg/sql/jobs_profiler_execution_details_test.go:275]: debugging detritus, deleted

File: pkg/sql/jobs_profiler_execution_details_test.go (hunk 3 above)
Comment (adityamaru) [reply pkg/sql/jobs_profiler_execution_details_test.go:243]: TIL, changed

**Review by stevendanna — APPROVED** [review]

**Comment by adityamaru** [comment]
TFTR!

bors r=stevendanna

**Pull request merged**